vowpalwabbit/ftrl.cc

```diff
@@ -252,54 +252,108 @@ void update_state_and_predict_pistol(ftrl& b, base_learner&, example& ec)
   ec.pred.scalar = GD::finalize_prediction(b.all->sd, b.all->logger, ec.partial_prediction);
 }
 
+template <bool privacy_activation>
 void update_after_prediction_proximal(ftrl& b, example& ec)
 {
   b.data.update = b.all->loss->first_derivative(b.all->sd, ec.pred.scalar, ec.l.simple.label) * ec.weight;
 
-  GD::foreach_feature<ftrl_update_data, inner_update_proximal>(*b.all, ec, b.data);
+  if (b.all->weights.sparse && privacy_activation)
+  {
+    b.all->weights.sparse_weights.set_tag(
+        hashall(ec.tag.begin(), ec.tag.size(), b.all->hash_seed) % b.all->feature_bitset_size);
+    GD::foreach_feature<ftrl_update_data, inner_update_proximal>(*b.all, ec, b.data);
+    b.all->weights.sparse_weights.unset_tag();
+  }
+  else if (!b.all->weights.sparse && privacy_activation)
+  {
+    b.all->weights.dense_weights.set_tag(
+        hashall(ec.tag.begin(), ec.tag.size(), b.all->hash_seed) % b.all->feature_bitset_size);
+    GD::foreach_feature<ftrl_update_data, inner_update_proximal>(*b.all, ec, b.data);
+    b.all->weights.dense_weights.unset_tag();
+  }
+  else
+  {
+    GD::foreach_feature<ftrl_update_data, inner_update_proximal>(*b.all, ec, b.data);
+  }
```

Comment:
This is a massive uptick in complexity when just doing an update. Can this be abstracted?